crates/client-api/src/auth.rs

```diff
@@ -37,95 +38,91 @@ impl authorization::Credentials for SpacetimeCreds {
     type Rejection = AuthorizationRejection;
     async fn from_request_parts(parts: &mut request::Parts, state: &S) -> Result<Self, Self::Rejection> {
-        match (
-            TypedHeader::from_request_parts(parts, state).await,
-            Query::<TokenQueryParam>::from_request_parts(parts, state).await,
-        ) {
-            (Ok(TypedHeader(headers::Authorization(creds @ SpacetimeCreds { .. }))), _) => {
-                let claims = creds
-                    .decode_token(state.public_key())
-                    .map_err(|e| AuthorizationRejection {
-                        reason: AuthorizationRejectionReason::Jwt(e.into_kind()),
-                    })?;
-                let auth = SpacetimeAuth {
-                    creds,
-                    identity: claims.identity,
-                };
-                Ok(Self { auth: Some(auth) })
-            }
-            (_, Ok(Query(query))) => {
-                let header =
-                    HeaderValue::from_str(&format!("Basic {}", query.token)).map_err(|_| AuthorizationRejection {
-                        reason: AuthorizationRejectionReason::MalformedTokenQueryString,
-                    })?;
-                let creds = SpacetimeCreds(authorization::Basic::decode(&header).ok_or(AuthorizationRejection {
-                    reason: AuthorizationRejectionReason::CantDecodeAuthorizationToken,
-                })?);
-                let claims = creds
-                    .decode_token(state.public_key())
-                    .map_err(|e| AuthorizationRejection {
-                        reason: AuthorizationRejectionReason::Jwt(e.into_kind()),
-                    })?;
-                let auth = SpacetimeAuth {
-                    creds,
-                    identity: claims.identity,
-                };
-                Ok(Self { auth: Some(auth) })
-            }
-            (Err(e), Err(_)) => match e.reason() {
-                // Leave it to handlers to decide on unauthorized requests.
-                TypedHeaderRejectionReason::Missing => Ok(Self { auth: None }),
-                _ => Err(AuthorizationRejection {
-                    reason: AuthorizationRejectionReason::Header(e),
-                }),
-            },
-        }
+        let Some(creds) = SpacetimeCreds::from_request_parts(parts)? else {
+            return Ok(Self { auth: None });
+        };
+        let claims = creds.decode_token(state.public_key())?;
+        let auth = SpacetimeAuth {
+            creds,
+            identity: claims.identity,
+        };
+        Ok(Self { auth: Some(auth) })
     }
 }
 
 /// A response by the API signifying that an authorization was rejected with the `reason` for this.
-pub struct AuthorizationRejection {
-    /// The reason the authorization was rejected.
-    reason: AuthorizationRejectionReason,
+#[derive(Debug, derive_more::From)]
+pub enum AuthorizationRejection {
+    Jwt(JwtError),
+    Header(headers::Error),
+    Required,
```

Comment:
this used to be `AuthorizationRejectionReason` below, but was moved up and refactored